### PR TITLE
Returning original index for dice-KD

### PR DESCRIPTION
**Pull request by soundarya98**
Signed-off-by: Soundarya Krishnan <soundaryak4898@gmail.com>
In response to issue #169 
- Made changes to show original indices of rows in the training data while displaying counterfactuals
- Added a test to ensure that this logic is being followed
- Fixed bug in dice_random when 0 cfs were being returned

### CODE REVIEW

File: dice_ml/explainer_interfaces/explainer_base.py [hunk 1]
```diff
@@ -392,14 +390,13 @@ def do_posthoc_sparsity_enhancement(self, final_cfs_sparse, query_instance, post
 
         cfs_preds_sparse = []
 
-        for cf_ix in range(len(final_cfs_sparse)):
-            current_pred = self.predict_fn_for_sparsity(
-                final_cfs_sparse.iloc[[cf_ix]][self.data_interface.feature_names])
+        for cf_ix in list(final_cfs_sparse.index):
+            current_pred = self.predict_fn_for_sparsity(final_cfs_sparse.loc[[cf_ix]][self.data_interface.feature_names])
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/explainer_base.py:394]: better to use `at` here for speed

File: dice_ml/explainer_interfaces/explainer_base.py [hunk 2]
```diff
@@ -392,14 +390,13 @@ def do_posthoc_sparsity_enhancement(self, final_cfs_sparse, query_instance, post
...
             for feature in features_sorted:
                 # current_pred = self.predict_fn_for_sparsity(final_cfs_sparse.iat[[cf_ix]][self.data_interface.feature_names])
                 # feat_ix = self.data_interface.continuous_feature_names.index(feature)
-                diff = query_instance[feature].iat[0] - final_cfs_sparse[feature].iat[cf_ix]
-                if abs(diff) <= quantiles[feature]:
+                diff = query_instance[feature].iat[0] - int(final_cfs_sparse.loc[[cf_ix]][feature])
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/explainer_base.py:398]: same comment. `at` is the analogue of `iat`, just like loc for iloc. `at` is much faster than `loc` when retrieving a single row/value, so let's use that everywhere.

File: dice_ml/explainer_interfaces/explainer_base.py [hunk 3]
```diff
@@ -408,7 +405,7 @@ def do_posthoc_sparsity_enhancement(self, final_cfs_sparse, query_instance, post
                         final_cfs_sparse = self.do_binary_search(
                             diff, decimal_prec, query_instance, cf_ix, feature, final_cfs_sparse, current_pred)
 
-            temp_preds = self.predict_fn_for_sparsity(final_cfs_sparse.iloc[[cf_ix]][self.data_interface.feature_names])
+            temp_preds = self.predict_fn_for_sparsity(final_cfs_sparse.loc[[cf_ix]][self.data_interface.feature_names])
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/explainer_base.py:408]: use `at`

File: dice_ml/explainer_interfaces/explainer_base.py [hunk 4]
```diff
@@ -424,28 +421,28 @@ def do_linear_search(self, diff, decimal_prec, query_instance, cf_ix, feature, f
         current_pred = current_pred_orig
         if self.model.model_type == ModelTypes.Classifier:
             while((abs(diff) > 10e-4) and (np.sign(diff*old_diff) > 0) and self.is_cf_valid(current_pred)):
-                old_val = final_cfs_sparse[feature].iat[cf_ix]
+                old_val = int(final_cfs_sparse.loc[[cf_ix]][feature])
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/explainer_base.py:424]: use `at`

**Review by amit-sharma — CHANGES_REQUESTED** [review]
Looks good, but need to change loc to at everywhere. `at` is much faster, see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html

File: dice_ml/diverse_counterfactuals.py [hunk 5]
```diff
@@ -114,7 +114,7 @@ def display_df(self, df, show_only_changes):
                         newdf[ix][jx] = '-'
                     else:
                         newdf[ix][jx] = str(newdf[ix][jx])
-            display(pd.DataFrame(newdf, columns=df.columns))  # works only in Jupyter notebook
```
Comment (gaugup) [inline dice_ml/diverse_counterfactuals.py:117]: How does the notebook output look like with this change?

Comment (soundarya98) [reply dice_ml/diverse_counterfactuals.py:117]: The index gets printed in the jupyter notebook

File: dice_ml/explainer_interfaces/explainer_base.py (hunk 1 above)
Comment (soundarya98) [reply dice_ml/explainer_interfaces/explainer_base.py:394]: We can't use `at` here since you can only query a single column using that

**Review by amit-sharma — APPROVED** [review]

**Pull request merged**